### PR TITLE
add TS types and tests

### LINKED ISSUE #28: Add TypeScript definitions
## 🚀 Feature Proposal

Add an index.d.ts with Typescript type definitions.

## Motivation

To use it on Typescript projects.

I can do a PR for this if you are ok with it.

### PR DESCRIPTION
**Pull request by ricardo-devis-agullo**
Closes #28 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: plugin.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Definitions by: ricardo-devis-agullo <https://github.com/ricardo-devis-agullo>
```
Comment (jsumners) [inline plugin.d.ts]: Do we really need this comment?

Comment (ricardo-devis-agullo) [reply plugin.d.ts]: Nope, I just copied from other fastify projects that had similar annotations. Can remove it.

Comment (mcollina) [reply plugin.d.ts]: Please do, thanks!

File: plugin.d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import fastify = require('fastify');
+
+import { Server, IncomingMessage, ServerResponse } from 'http';
+
+declare const fastifyBearerAuth: fastify.Plugin<
+  Server,
+  IncomingMessage,
+  ServerResponse,
```
Comment (mcollina) [inline plugin.d.ts]: Can you please add support for http2 here as well? Thanks!

File: plugin.d.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  {
+    keys: Set<string> | string;
```
Comment (orblazer) [inline plugin.d.ts]: The type is wrong, the correct type : Set<string> | string[]

Comment (orblazer) [inline plugin.d.ts]: The type is wrong, the correct type is : `Set<string> | string[]`

**Comment by Ethan-Arrowood** [comment]
Closing in favor of v3 update

**Pull request closed without merging**